marimo/_ast/codegen.py

```diff
@@ -174,59 +263,110 @@ def generate_filecontents(
+            )
 
     unshadowed_builtins = set(builtins.__dict__.keys()) - defs
-    fndefs: list[str] = []
-
-    # Update old internal cell names to the new ones
-    for idx, name in enumerate(names):
-        if name == "__":
-            names[idx] = DEFAULT_CELL_NAME
-
-    for data, name in zip(cell_data, names):
-        if isinstance(data, CellImpl):
-            if toplevel_fn and data.is_toplevel_acceptable:
-                fndefs.append(to_top_functiondef(data))
-            else:
-                fndefs.append(to_functiondef(data, name, unshadowed_builtins))
-        else:
-            fndefs.append(
-                generate_unparsable_cell(
-                    code=data[0], config=data[1], name=name
-                )
+    allowed_refs = unshadowed_builtins | toplevel_imports
+
+    for idx, cell in enumerate(definition_stubs):
+        # We actually don't care about the graph, we just want to see if we can
+        # render the top-level functions without a name error.
+        # Let graph issues be delegated to the runtime.
+        if cell and toplevel_fn and cell.is_toplevel_acceptable(allowed_refs):
+            definitions[idx] = to_top_functiondef(cell, allowed_refs)
+            definition_stubs[idx] = None
+            # Order does matter since feasibly, an app.function could be a
+            # decorator for another.
+            allowed_refs.add(names[idx])
+
+    # Let's hide the new behavior for now.
+    # Removing the toplevel_fn check may produce a bit of churn,
+    # so let's release the serialization changes all together.
+    if not toplevel_fn:
+        allowed_refs = unshadowed_builtins
+        used_refs = None
+
+    for idx, cell in enumerate(definition_stubs):
+        if cell:
+            definitions[idx] = to_functiondef(
+                cell, names[idx], allowed_refs, used_refs, fn="cell"
             )
 
-    filecontents = (
-        "import marimo"
-        + "\n\n"
-        + f'__generated_with = "{__version__}"'
-        + "\n"
-        + generate_app_constructor(config)
-        + "\n\n\n"
-        + "\n\n\n".join(fndefs)
-        + "\n\n\n"
-        + 'if __name__ == "__main__":'
-        + "\n"
-        + indent_text("app.run()")
+    assert all(isinstance(d, str) for d in definitions)
+    cell_blocks: List[str] = cast(List[str], definitions)
+    if not toplevel_fn:
+        return _classic_export(cell_blocks, header_comments, config)
+
+    filecontents = []
+    if header_comments is not None:
+        if not toplevel_fn:
+            filecontents = [header_comments, ""]
+        elif header_comments := header_comments.rstrip():
+            filecontents = [header_comments, "", NOTICE, ""]
+        else:
+            # If the header comments are empty, we may still need to contain the
+            # notice to separate the imports from the rest of the code.
+            filecontents = [NOTICE, ""]
+
+    filecontents.append("\n\n".join(import_blocks))
```

Comment:
Should imports be added unconditionally, as you've written, or should imports only be added if they are used in top-level functions?

One thought, if imports are added to the top of the file unconditionally, perhaps we should remove their corresponding defs from cell signatures, so that code completion in editors works better. However, maybe the right thing to do is just bite the bullet and write editor plugins / an LSP-like thing that handle completions for marimo notebook files, in which case my suggestion here is moot.

Reply:
Also, can we `ruff format` the import section?

Reply:
> perhaps we should remove their corresponding defs from cell signatures, so that code completion in editors works better

Yep, this PR already does this

> Also, can we ruff format the import section?

Yes, I'm leaning towards removing the statement block, stripping comments and formatting the imports.
Thoughts?

Reply:
I think that sounds good ... also see my response to your import guard idea.